pkg/reconcile/runner.go

```diff
@@ -0,0 +1,158 @@
+type Loop interface {
+	RunReconciler(ctx context.Context, reconciler Reconciler, predicates ...predicate.Predicate) error
+}
+
+type runner struct {
+	name     string
+	mgr      manager.Manager
+	resource ezkube.Object
+}
+
+func NewLoop(name string, mgr manager.Manager, resource ezkube.Object) *runner {
+	return &runner{name: name, mgr: mgr, resource: resource}
+}
+
+type runnerReconciler struct {
+	ctx        context.Context
+	mgr        manager.Manager
+	resource   ezkube.Object
+	logger     logr.Logger
+	reconciler Reconciler
+}
+
+func (ec *runner) RunReconciler(ctx context.Context, reconciler Reconciler, predicates ...predicate.Predicate) error {
+	gvk, err := apiutil.GVKForObject(ec.resource, ec.mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+	rec := &runnerReconciler{
+		logger:     log.Log.WithName("event-controller").WithValues("kind", gvk).WithName(ec.name),
+		mgr:        ec.mgr,
+		resource:   ec.resource,
+		reconciler: reconciler,
+	}
+
+	ctl, err := controller.New(ec.name, ec.mgr, controller.Options{
+		Reconciler: rec,
+	})
+	if err != nil {
+		return err
+	}
+
+	// send us watch events
+	if err := ctl.Watch(&source.Kind{Type: ec.resource}, &handler.EnqueueRequestForObject{}, predicates...); err != nil {
+		return err
+	}
+
+	rec.logger.V(1).Info("waiting for cache sync...")
+	if synced := ec.mgr.GetCache().WaitForCacheSync(ctx.Done()); !synced {
+		return errors.Errorf("waiting for cache sync failed")
+	}
+
+	return nil
+}
+
+func (ec *runnerReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := ec.logger.WithValues("event", request)
+	logger.V(2).Info("handling event", "event", request)
+
+	// get the object from our cache
+	restClient := ezkube.NewRestClient(ec.mgr)
+
+	obj := ec.resource.DeepCopyObject().(ezkube.Object)
+	obj.SetName(request.Name)
+	obj.SetNamespace(request.Namespace)
+	if err := restClient.Get(ec.ctx, obj); err != nil {
+		logger.Error(err, "unable to fetch %T %v", obj, request)
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// if the handler is a finalizer, check if we need to finalize
+	if finalizer, ok := ec.reconciler.(FinalizingReconciler); ok {
+		finalizers := obj.GetFinalizers()
+		finalizerName := finalizer.FinalizerName()
+		if obj.GetDeletionTimestamp().IsZero() {
+			// The object is not being deleted, so if it does not have our finalizer,
+			// then lets add the finalizer and update the object. This is equivalent to
+			// registering our finalizer.
+
+			if !utils.ContainsString(finalizers, finalizerName) {
+				obj.SetFinalizers(append(
+					finalizers,
+					finalizerName,
+				))
+				if err := restClient.Update(context.Background(), obj); err != nil {
+					return reconcile.Result{}, err
+				}
+			}
```

Comment:
Do we want to return at the end of this `if` block. We most likely do not want to finalize if the DeletionTimestamp `IsZero()`